### PR TITLE
fix(db) Allow RDS + null databases to coexist

### LINKED ISSUE #122: Drivers::Framework::Base#link_sqlite_database does not play nice with RDS present + null DB adapters
(Opening this issue to reference in a PR.)

This is a bit of a corner case, but I have it in my environment where we deploy four apps: three with null DB adapters and one that uses RDS.

In the current implementation of [every_enabled_rds](https://github.com/ajgon/opsworks_ruby/blob/master/libraries/helpers.rb#L92) the code assumes that the list of interesting databases are *either* those provided by the AWS RDS search, or the one constructed based on the node engine. However in our case both might be true! Some apps might want RDS instances while others might want a configuration constructed from node data.

Because of this implementation of `every_enabled_rds`, we can unexpectedly fall to a line in [`Drivers::Framework::Base#database_url`](https://github.com/ajgon/opsworks_ruby/blob/master/libraries/drivers_framework_base.rb#L81) where we incorrectly conclude that the app wants to use a sqlite DB when in fact it wants to use the `null` DB adapter. Because of the absence of a `db` subdirectory in the `release_path`, the cookbook explodes trying to symlink the unwanted sqlite DB into place.

Adding a `db` directory to my apps is not an option, either in the git repo (I can't justify it) or in the deploy process (these hooks fire too early in the deploy resource for me to create the directory), so really I just want to fix the logic to avoid the problem.

The general shape of the solution should be:
- `every_enabled_rds` contains both the active RDS instances *and* the node-driven Drivers::Db::* object that is appropriate for the `app` in question.
- Drivers::Framework::Base stops trying to sorta-kinda implement any sqlite-as-default behavior
- The sqlite deploy hooks move into the Drivers::Db::Sqlite module where they belong.
- Drivers::Db::Factory.build *does* implement sqlite-as-default, but only if you don't specify any database adapter.

### PR DESCRIPTION
**Pull request by nickmarden**
Fixes #122

### CODE REVIEW

File: libraries/drivers_db_sqlite.rb [hunk 1]
```diff
@@ -7,6 +7,21 @@ class Sqlite < Base
...
+
+      def validate_node_engine
+        unless node_engine
+          framework = Drivers::Framework::Factory.build(context, app)
+          deploy = context.node['deploy']
+          deploy[app['shortname']][driver_type] ||= {}
+          deploy[app['shortname']][driver_type][:database] = "db/#{app['shortname']}_#{framework.deploy_env}.sqlite"
+          context.node['deploy'] = deploy
+        end
+        :node_engine
```
Comment (ajgon) [inline libraries/drivers_db_sqlite.rb]: Please use guard clause here, so instead this big unless:

```ruby
return :node_engine if node_engine
framework = ....
```

File: libraries/drivers_db_sqlite.rb [hunk 2]
```diff
@@ -16,6 +31,32 @@ def out
...
+          not_if { ::File.exist?(shared_directory_path) }
+        end
+
+        context.file File.join(deploy_to, 'shared', relative_db_path) do
+          action :create
+          not_if { ::File.exist?(File.join(deploy_to, 'shared', relative_db_path)) }
+        end
+
+        context.link File.join(release_path, relative_db_path) do
+          to db_path
```
Comment (ajgon) [inline libraries/drivers_db_sqlite.rb]: `db_path` is missing and tests are failing:
```
 [2017-10-20T05:03:35+00:00] ERROR: deploy[dummy_project] (opsworks_ruby::deploy line 22) had an error: NoMethodError: undefined method `db_path' for Chef::Resource::Link
```

Comment (nickmarden) [reply libraries/drivers_db_sqlite.rb]: Oh that's embarrassing. Sorry for pushing a broken PR :-/

Fix coming up.

File: libraries/drivers_db_sqlite.rb (hunk 1 above)
Comment (nickmarden) [reply libraries/drivers_db_sqlite.rb]: The only thing is, the `return :node_engine` happens in _all_ cases, so it's not a guard clause per se. But I'll clean it up to make it more readable.

**Comment by nickmarden** [comment]
OK sorry for the stupidity @ajgon. Specs and integration tests are working now.

**Pull request merged**